src/svd/util.py

```diff
@@ -0,0 +1,386 @@
+        # svada does not have a way to check written status at this level currently
+        if (
+            self.content_status == BuildSelector.ContentStatus.MODIFIED
+            and not field.modified
+        ):
+            return False
+
+        return True
+
+
+class DeviceBuilder:
+    """Used to populate peripheral registers and output the register contents in various formats."""
+
+    def __init__(
+        self, device: svd.Device, enforce_svd_constraints: bool = True
+    ) -> None:
+        self._device = device
+        self._enforce_svd_constraints = enforce_svd_constraints
+        self._written_peripherals = {}
+        self._cached_device_periph_map = []
+        self._cached_periph_reg_maps = {}
+
+    @property
+    def device(self) -> svd.Device:
+        """The device structure the builder was initialized with."""
+        return self._device
+
+    @property
+    def written_peripherals(self) -> list[svd.Peripheral]:
+        """The list of peripherals that have been written to as part of the API calls."""
+        return list(self._written_peripherals.values())
+
+    def build_bytes(self, selector: BuildSelector = BuildSelector()) -> bytearray:
+        """Encode device content as bytes.
+
+        :param selector: selected parts of the device.
+        :returns: content bytes.
+        """
+        memory = self.build_memory(selector)
+        out = bytearray()
+
+        for (addr_a, value_a), (addr_b, value_b) in pairwise(memory.items()):
+            if not out:
+                out.append(value_a)
+
+            num_empty = addr_b - addr_a - 1
+            if num_empty > 0:
+                # TODO: 0 may not always be valid
+                out.extend([0] * num_empty)
+
+            out.append(value_b)
+
+        return out
+
+    def build_memory(self, selector: BuildSelector = BuildSelector()) -> dict[int, int]:
+        """Encode device content as a mapping between address and value.
+
+        :param selector: selected parts of the device.
+        :returns: content memory map.
+        """
+        memory = {}
+
+        for peripheral in self._device.values():
+            if not selector.is_periph_selected(peripheral):
+                continue
+
+            if selector.content_status == BuildSelector.ContentStatus.MODIFIED:
+                # There's no good way to determine this in svada at the moment
+                periph_modified_filter = set()
+                for reg in peripheral.register_iter(leaf_only=True):
+                    if reg.modified:
+                        periph_modified_filter.update(reg.address_range)
+            else:
+                periph_modified_filter = None
+
+            memory_iter = peripheral.memory_iter(
+                absolute_addresses=True,
+                written_only=(
+                    selector.content_status == BuildSelector.ContentStatus.WRITTEN
+                ),
+            )
+
+            for addr, val in memory_iter:
```

Comment:
Was `is_addr_selected()` intended for this? It's currently unused, hence address ranges don't work here.